File: AprilTagScanner/Services/ScanSession.cs
using AprilTagScanner.Models;

namespace AprilTagScanner.Services;

public sealed class ScanSession
{
    private readonly object _lock = new();
    private readonly HashSet<TagKey> _visibleNow = [];
    private readonly Dictionary<TagKey, int> _missCounts = [];
    private readonly List<TagRecord> _records = [];
    private readonly HashSet<TagKey> _duplicates = [];

    public IReadOnlyList<TagRecord> Records
    {
        get { lock (_lock) return _records.ToList(); }
    }

    public IReadOnlySet<TagKey> Duplicates
    {
        get { lock (_lock) return _duplicates.ToHashSet(); }
    }

    public void Reset()
    {
        lock (_lock)
        {
            _records.Clear();
            _duplicates.Clear();
            _visibleNow.Clear();
            _missCounts.Clear();
        }
    }

    public void ClearTracking()
    {
        lock (_lock)
        {
            _visibleNow.Clear();
            _missCounts.Clear();
        }
    }

    public bool ShouldRecord(TagKey key, out bool isDuplicate)
    {
        lock (_lock)
            return ShouldRecordUnlocked(key, out isDuplicate);
    }

    private bool ShouldRecordUnlocked(TagKey key, out bool isDuplicate)
    {
        isDuplicate = false;
        var lastPos = -1;
        for (var i = 0; i < _records.Count; i++)
        {
            if (new TagKey(_records[i].Family, _records[i].Id) == key)
                lastPos = i;
        }

        if (lastPos < 0)
            return true;

        for (var i = lastPos + 1; i < _records.Count; i++)
        {
            if (new TagKey(_records[i].Family, _records[i].Id) != key)
            {
                isDuplicate = true;
                return true;
            }
        }

        return false;
    }

    public TagRecord? TryAppend(DetectedTag tag)
    {
        lock (_lock)
            return TryAppendUnlocked(tag);
    }

    private TagRecord? TryAppendUnlocked(DetectedTag tag)
    {
        if (!ShouldRecordUnlocked(tag.Key, out var isDup))
            return null;

        var record = new TagRecord
        {
            Family = tag.Family,
            Id = tag.Id,
            Duplicate = isDup,
        };
        if (isDup)
            _duplicates.Add(tag.Key);
        _records.Add(record);
        return record;
    }

    public IReadOnlyList<TagRecord> ProcessDetections(IReadOnlyList<DetectedTag> tags, int missLimit)
    {
        lock (_lock)
        {
            var detected = new Dictionary<TagKey, DetectedTag>();
            foreach (var tag in tags)
                detected.TryAdd(tag.Key, tag);

            var appended = new List<TagRecord>();

            foreach (var key in _visibleNow.ToList())
            {
                if (detected.ContainsKey(key))
                {
                    _missCounts[key] = 0;
                    continue;
                }

                var missed = _missCounts.GetValueOrDefault(key) + 1;
                _missCounts[key] = missed;
                if (missed >= missLimit)
                {
                    _visibleNow.Remove(key);
                    _missCounts.Remove(key);
                }
            }

            var newKeys = detected.Keys.Except(_visibleNow)
                .OrderBy(k => detected[k].Center.Y)
                .ThenBy(k => detected[k].Center.X)
                .ToList();

            foreach (var key in newKeys)
            {
                var record = TryAppendUnlocked(detected[key]);
                if (record != null)
                    appended.Add(record);
                _visibleNow.Add(key);
                _missCounts[key] = 0;
            }

            return appended;
        }
    }
}

File: AprilTagScannerAndroid/app/src/main/kotlin/com/scanapriltag/services/ScanSession.kt
package com.scanapriltag.services

import com.scanapriltag.models.DetectedTag
import com.scanapriltag.models.TagKey
import com.scanapriltag.models.TagRecord

/**
 * Session state is mutated from the camera analysis thread and read from main
 * (UI / export). All public access goes through [lock].
 */
class ScanSession {
    private val lock = Any()

    private val visibleNow = mutableSetOf<TagKey>()
    private val missCounts = mutableMapOf<TagKey, Int>()
    private val recordsMut = mutableListOf<TagRecord>()
    private val duplicatesMut = mutableSetOf<TagKey>()

    /** Immutable snapshot — safe to iterate off the analysis thread. */
    val records: List<TagRecord>
        get() = synchronized(lock) { recordsMut.toList() }

    /** Immutable snapshot for overlay / export. */
    val duplicates: Set<TagKey>
        get() = synchronized(lock) { duplicatesMut.toSet() }

    fun reset() {
        synchronized(lock) {
            recordsMut.clear()
            duplicatesMut.clear()
            visibleNow.clear()
            missCounts.clear()
        }
    }

    fun clearTracking() {
        synchronized(lock) {
            visibleNow.clear()
            missCounts.clear()
        }
    }

    fun shouldRecord(key: TagKey): Pair<Boolean, Boolean> {
        synchronized(lock) {
            return shouldRecordLocked(key)
        }
    }

    fun tryAppend(tag: DetectedTag): TagRecord? {
        synchronized(lock) {
            return tryAppendLocked(tag)
        }
    }

    fun processDetections(tags: List<DetectedTag>, missLimit: Int): List<TagRecord> {
        synchronized(lock) {
            val detected = tags.associateBy { it.key }
            val appended = mutableListOf<TagRecord>()

            for (key in visibleNow.toList()) {
                if (detected.containsKey(key)) {
                    missCounts[key] = 0
                    continue
                }

                val missed = (missCounts[key] ?: 0) + 1
                missCounts[key] = missed
                if (missed >= missLimit) {
                    visibleNow.remove(key)
                    missCounts.remove(key)
                }
            }

            val newKeys = detected.keys
                .filter { it !in visibleNow }
                .sortedWith(
                    compareBy<TagKey> { detected[it]!!.center.y }
                        .thenBy { detected[it]!!.center.x },
                )

            for (key in newKeys) {
                val record = tryAppendLocked(detected[key]!!)
                if (record != null) appended.add(record)
                visibleNow.add(key)
                missCounts[key] = 0
            }

            return appended
        }
    }

    private fun shouldRecordLocked(key: TagKey): Pair<Boolean, Boolean> {
        val keys = recordsMut.map { TagKey(it.family, it.id) }
        if (!keys.contains(key)) return true to false

        val lastPos = keys.indexOfLast { it == key }
        val otherSince = keys.drop(lastPos + 1).any { it != key }
        if (!otherSince) return false to false

        return true to true
    }

    private fun tryAppendLocked(tag: DetectedTag): TagRecord? {
        val (should, isDup) = shouldRecordLocked(tag.key)
        if (!should) return null

        val record = TagRecord(tag.family, tag.id, isDup)
        if (isDup) duplicatesMut.add(tag.key)
        recordsMut.add(record)
        return record
    }
}
